dds/src/implementation/utils/timer_factory.rs

```diff
@@ -0,0 +1,313 @@
+    }
+}
+
+pub struct TimerProvider {
+    instance_timers: HashMap<InstanceHandle, Timer>,
+    timer_condvar: DdsCondvar,
+}
+
+impl TimerProvider {
+    fn new(timer_condvar: DdsCondvar) -> Self {
+        Self {
+            instance_timers: HashMap::new(),
+            timer_condvar,
+        }
+    }
+
+    pub fn start_timer(
+        &mut self,
+        duration: std::time::Duration,
+        id: InstanceHandle,
+        func: impl Fn() + 'static + Send + Sync,
+    ) {
+        self.instance_timers.insert(id, Timer::new(duration, func));
+        self.timer_condvar.notify_all();
+    }
+
+    pub fn cancel_timers(&mut self) {
+        self.instance_timers.clear();
+    }
+}
+
+pub struct TimerFactory {
+    timer_provider_list: DdsShared<DdsRwLock<Vec<DdsShared<DdsRwLock<TimerProvider>>>>>,
+    timer_condvar: DdsCondvar,
+    thread_handle: Option<JoinHandle<()>>,
+    should_stop: DdsShared<std::sync::atomic::AtomicBool>,
+}
+
+impl Default for TimerFactory {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TimerFactory {
+    pub fn new() -> Self {
+        let timer_provider_list = DdsShared::new(DdsRwLock::new(Vec::<
+            DdsShared<DdsRwLock<TimerProvider>>,
+        >::new()));
+        let should_stop = DdsShared::new(AtomicBool::new(false));
+        let timer_condvar = DdsCondvar::new();
+
+        let timer_provider_list_clone = timer_provider_list.clone();
+        let should_stop_clone = should_stop.clone();
+        let timer_condvar_clone = timer_condvar.clone();
+
+        let thread_handle = std::thread::spawn(move || loop {
+            if should_stop_clone.load(Ordering::Relaxed) {
+                break;
+            }
+
+            for timer_provider in timer_provider_list_clone.read_lock().iter() {
+                for v in timer_provider.write_lock().instance_timers.values_mut() {
+                    if v.is_elapsed() {
+                        v.reset();
+                        (v.func)()
+                    }
+                }
+            }
+
+            let min_remaining_duration = timer_provider_list_clone
+                .read_lock()
+                .iter()
+                .map(|x| {
+                    x.read_lock()
+                        .instance_timers
+                        .values()
+                        .map(|x| x.remaining_duration())
+                        .min()
+                })
+                .min();
+
+            if let Some(Some(d)) = min_remaining_duration {
+                timer_condvar_clone.wait_timeout(d.into()).ok();
+            } else {
+                timer_condvar_clone.wait().ok();
+            }
+        });
+
+        Self {
+            timer_provider_list,
+            timer_condvar,
+            thread_handle: Some(thread_handle),
+            should_stop,
+        }
+    }
+
+    pub fn create_timer_provider(&self) -> DdsShared<DdsRwLock<TimerProvider>> {

```

Comment:
Since the TimerFactory manufactures a very specific timer here, namely one that receives an InstganceHandle as argument, it might be called more like that specific type. The TimerFactory may then also manufature other timer types. 
maybe: create_per_instance_timer() and for a timer that is periodically called type create_periodic_timer()